Hash maps

wclodius2 opened the pull request:
<!--
Thank you for contributing to stdlib.
To help us get your pull request merged more quickly, please consider reviewing any of the already open pull requests.
-->

Now that hash functions have been added to the standard library, it is time to add hash maps. Currently this PR only includes the documentation, `doc/specs/stdlib_hash_maps.md`, giving the proposed API. The proposed API has three module files: `stdlib_key_data_wrapper.f90`, which provides wrappers for the key and data components of the entries and some of the hash function; `stdlib_chaining_hash_map.f90`, which implements a chaining hash map with linked lists; and `stdlib_open_hash_maps.f90`, which implements a linear addressing open hashing map. After the (modified) API  has been approved by two people I will add the required module files.

awvwgk reviewed (commented):
Three general questions:

1. Why represent the key / other type as array of `int8` rather than as `class(*), allocatable? At least for the other type the latter representation would be more flexible without the need for memory manipulations with `transfer`
2. Would the definition of a container type be useful? This one could be used for all data structures (*e.g.* lists).
3. Should the procedures for the maps be type bound? First this makes imports simpler and might also allow to define a common base class.

awvwgk commented:
Regarding the naming of the modules I would propose to adjust the naming to be more concise:

- `stdlib_key_data_wrapper` -> `stdlib_container`
- `stdlib_chaining_hash_map` -> `stdlib_hashmap_chaining`
- `stdlib_open_hash_map` -> `stdlib_hashmap_open`

If we can define an abstract base class, `stdlib_map_class` (for both ordered and unordered maps) or `stdlib_hashmap_class` (for unordered maps only) module might be a good name.

wclodius2 commented:
@awvwgk 
1. I have never used `CLASS(*)`. I assume I can ask for the underlying type using `select(type)`. Does it preserve array information so I can ask rank and array size array size? Are rank independent procedure arguments widely implemented?
2. I don't know if it will be useful to define a container type beyond what I define as the `open_map_entry_type` and `chaining_map_entry_type`. Would it be intended to replace the `key` and `other` components of those two types?
3. I had started to define a parent type (in a module `stdlib_hash_maps`) to the `open_hash_map_type` and `chaining_hash_map_type`. It had some `non_overridable` type bound procedures and a large number of `deferred` type bound procedures. Implementing it fully would require decisions as to whether we want the use of inheritance, and whether the detailed implementation of the subtypes would be in modules or sub-modules.

wclodius2 commented:
@awvwgk the proposed module names seem to me (and @milancurcic) better than what I currently have so I will change the modules and their files to what you suggest.

wclodius2 commented:
In thinking further I am inclined tho rename `stdlib_key_data_wrapper` to `stdlib_hashmap_container`.

gareth-nx commented:
> 1. have never used `CLASS(*)`. I assume I can ask for the underlying type using `select(type)`. Does it preserve array information so I can ask rank and array size array size? Are rank independent procedure arguments widely implemented?

I believe that `CLASS(*)` objects cannot be arrays. If you want it to hold an array then you have to create a derived type that holds the array.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3088 @@
+The hash codes, in turn, can be mapped to a smaller set of integers,
+that can be used as an index, termed a hash index, to a rank one
+array, often termed a hash table or hash map.
+This document discusses the hash maps in the library.
```
Here I think the text initially gives too much of an introduction to hash codes, outside of the context of hash maps. It might be better to start out explaining what a hash map is (in sufficiently vague terms). 

Not sure what the wording should be. Perhaps: "A hash map is a data-structure that maps keys to values. Such data structures occur in many languages, for example the python `dict`, the C++ `map`, ...".

wclodius2 replied:
Leet me see what I can come up with.

wclodius2 replied:
I have revised that paragraph.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3088 @@
+map noted more for its diagnostics than its performance. Finally the
+module, `stdlib_open_hash_map` defines a datatype,
+`open_hash_map_type`, implementing a simple open addressing hash
+map noted more for its diagnostics than its performance.
```
So both of the hash maps are `noted more for its diagnostics than its performance` ? That's OK, just want to double check that is the intention.

wclodius2 replied:
I think the performance is not baad. Generating the information used in the diagnostics is, I suspect, a small amount of overhead, but it is an overhead other tables will lack.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3088 @@
+the ratio of the number of hash map probes to the number of subroutine 
+calls.
+The maps make extensive use of pointers internally, but a private
+finalization subroutine avoids memory leaks.
```
Does this rely on the compiler to automatically call the finalization subroutine when a derived type goes out of scope? Last time I checked (albeit not recently) gfortran still didn't do this.

wclodius2 replied:
Yes it relies on this to avoid memory leaks, though the largest amount of memory will be in arrays and not in the linked list used to avoid memory allocations. I don't think there are memory leaks with my most recent code compiled using the latest version of `gfortran`, though my testing was very indirect.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3088 @@
+* `copy_key( key_in, key_out )` - Copies the contents of the key,
+  key_in, to the key, key_out.
+
+* `get( key, value )` - extracts the content of key into value.
```
Not immediately clear to me how this differs from `copy_key`

gareth-nx replied:
I see the details are explained below -- some minor edits here should make the difference clear.

wclodius2 replied:
`copy_key` copies the internals of one key to the internals of another key. `get` copies the contents of a key or other data into an allocatable default character string or `int8` array. I have improved the documentation of this in `stdlib_hash_maps.md`.

wclodius2 replied:
I added a few more details and pushed the result.

awvwgk reviewed (commented):
Another question, why is the index `inmap` part of the public API? Shouldn't this be an implementation detail of the map type and hidden from the user?

Is there a drawback to reference entries sorely with a key? To retain a reference we could have a pointer to the data returned, if we want to safe redundant queries on the map.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3088 @@
+* `copy_other( other_in, other_out )` - Copies the contents of the
+  other data in, other_in, to the other data, other_out.
+
+* `get( other, value )` - extracts the content of other into value.
```
Not clear to me how this differs from `copy_other`

wclodius2 replied:
See the above discussion for `copy_key` and `get`.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
````diff
@@ -0,0 +1,3088 @@
+```
+
+
+#### `FIBONACCI_HASH` - maps an integer to a smaller number of bits
````
Is this different to the version in `stdlib_hash_32bit`?

wclodius2 replied:
It is  just giving access to the version in `stdlib_hash_32bi`.

jvdp1 replied:
If it is the case, then i suggest to to repeat the specs here. A link to the page of `Fibonacci_hash` should be enough,

wclodius2 replied:
Do you mean "then i suggest not to repeat the specs here."?

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3088 @@
+      do i=1, 15
+          value(i) = i
+      end do
+      call set( key, value )
```
Should `key` be `key_in` ?

wclodius2 replied:
Yes. Fixed.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
````diff
@@ -0,0 +1,3088 @@
+```fortran
+    program demo_copy_key
+      use stdlib_32_bit_key_data_wrapper, only: &
+          copy_key, key_test, key_type
````
It looks like `key_test` is a test for equality? In that case consider if it would be better named `equal_keys`.

wclodius2 replied:
> > The current use of `inmap` is about a factor if two more efficient than using `key` directly for modifications of the `other` data associated with the `key`. For initially entering an `entry`, testing for existence of the `key` in the `map` , and deleting an `entry` from the `map` there is not much difference in the efficiency between using `inmap` and using `key` directly.
> 
> If querying the key twice is a performance issue, could you return a `pointer` to other rather than its index `inmap`?

I would not return a pointer to internal data. It can cause a memory leak or other problems, and would require making more of the internal `targets` or `pointers`, making the structure more fragile. It is a performance issue for a lesser used aspect of hash maps, and I would rather make a localized performance degradation than use pointers.

wclodius2 replied:
I don't know why this comment appeared here and not on the tail of the comments.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
````diff
@@ -0,0 +1,3088 @@
+```
+
+
+#### `HASHER_FUN`- serves aa function prototype.
````
typo

wclodius2 replied:
Fixed several typos in the documentation for `hasher_fun`.

gareth-nx commented on doc/specs/stdlib_hash_maps.md:
````diff
@@ -0,0 +1,3088 @@
+    end program demo_hasher_fun
+```
+
+#### `key_test` - Compares two keys for equality
````
As above, consider renaming to `equal_keys` or `keys_equal` or similar.

gareth-nx reviewed (commented):
Thanks, this is really important for `stdlib`. I don't have expertise in these data structures, but made a few minor comments.

wclodius2 commented:
> 
> I believe that `CLASS(*)` objects cannot be arrays. If you want it to hold an array then you have to create a derived type that holds the array.

Then I believe `CLASS(*)` will be of very limited use.

wclodius2 commented:
In thinking further I will rename `stdlib_key_data_wrapper` to `stdlib_hashmap_wrappers`.

wclodius2 commented:
@awvwgk I have changed the module names and revised the first paragraph to focus more on hash maps than hash functions., and have pushed the revised document.

wclodius2 commented:
> Another question, why is the index `inmap` part of the public API? Shouldn't this be an implementation detail of the map type and hidden from the user?
> 
> Is there a drawback to reference entries sorely with a key? To retain a reference we could have a pointer to the data returned, if we want to safe redundant queries on the map.

FWIW I have slightly improved the documentation for `inmap` in the current `stdlib_hash_maps.md`. The current use of `inmate` is about a factor if two more efficient than using `key` directly for modifications of the `other` data associated with the `key`. For initially entering an `entry`, testing for existence of the `key` in the `map` , and deleting an `entry` from the `map` there is not much difference in the efficiency between using `inmap` and using `key` directly. The use of `inmap` as an argument could be replaced by `key` and using `in_map` in the code of the procedure to generate `inmap` from the `key`.

awvwgk commented:
> The current use of `inmap` is about a factor if two more efficient than using `key` directly for modifications of the `other` data associated with the `key`. For initially entering an `entry`, testing for existence of the `key` in the `map` , and deleting an `entry` from the `map` there is not much difference in the efficiency between using `inmap` and using `key` directly.

If querying the key twice is a performance issue, could you return a `pointer` to other rather than its index `inmap`?

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+`other_type`, and associated procedures, for storing and manipulating
+keys and their associated data.
+
+### The `stdlib_hashmap_wrappers` constant, `INT_HASH`
```
```suggestion
### The `stdlib_hashmap_wrappers` constant, `int_hash`
```

wclodius2 replied:
I have tried to change all upper case names to lower case. Let me know if I missed any.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+### The `stdlib_hashmap_wrappers` constant, `INT_HASH`
+
+The constant `INT_HASH` is used to define the integer kind value for
+the returned hash codes and variables used to access them. It
+currently has the value, `INT32`.
```
```suggestion
The constant `int_hash` is used to define the integer kind value for
the returned hash codes and variables used to access them. It
currently has the value, `int32`.
```

wclodius2 replied:
See above,

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+represent keys and associated data stored in a module, but is also, a
+wrapper for the `stdlib_hash_32bit` module. It allows
+direct access to the `stdlib_hash_32bit` procedures:
+`fibonacci_hash`, `fnv_1_hasher`, `fnv_1a_hasher`; and provides
```
```suggestion
`fibonacci_hasher`, `fnv_1_hasher`, `fnv_1a_hasher`; and provides
```

wclodius2 replied:
No it should be `fibonacci_hash`

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+### Specifications of the `stdlib_hashmap_wrappers` procedures
+
+#### `copy_key` - Returns a copy of the key
```
could it overload "="?

wclodius2 replied:
Yes it could. It would require exchanging argument locations and changing all references to `copy_key` in the code. The same could be done for `copy_other`.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Result value
+
+The result is `.TRUE.` if the keys are equal, otherwise `.FALSS`.
```
```suggestion
The result is `.true.` if the keys are equal, otherwise `.false.`.
```

wclodius2 replied:
see above

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+is an `intent(in)` argument.
+
+`value`: shall be an allocatable default character string variable, or
+an allocatable vector variable of type integer and kind `INT8`. It is
```
```suggestion
an allocatable vector variable of type integer and kind `int8`. It is
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+##### Description
+
+Serves as a prototype for hashing functions with a single, `key`,
+argument returning an `INT32` hash value.
```
```suggestion
argument returning an `int32` hash value.
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Argument
+
+`key`: Shall be a rank one array expression of type `INTEGER(INT8)`.
```
```suggestion
`key`: Shall be a rank one array expression of type `integer(int8)`.
```

wclodius2 replied:
> Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Result character
+
+The result is a scalar integer of kind `INT32`.
```
```suggestion
The result is a scalar integer of kind `int32`.
```

wclodius2 replied:
Done

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Result character
+
+The result is a scalar integer of kind `INT32`.
```
```suggestion
The result is a scalar integer of kind `int32`.
```

wclodius2 replied:
Done

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+As a result it should give fair performance for typical hash map
+applications.
+This code passes the SMHasher tests.
+As a result it should give good performance for typical hash map
+applications.
```
Repeated sentence

```suggestion
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+As a result it should give fair performance for typical hash map
+applications.
+This code passes the SMHasher tests.
+As a result it should give good performance for typical hash map
+applications.
```
```suggestion
```

wclodius2 replied:
Done

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+As a result it should give reasonable performance for typical hash
+table applications.
+This code passes the SMHasher tests.
+As a result it should give good performance for typical hash map
+applications.
```
```suggestion
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+also defines a public data type and associated procedures and
+constants that implement a simple hash map using
+separate chaining hashing. The derived type is
+`chaining_hash_map_type`. It provides 
```
```suggestion
`chaining_hashmap_type`. It provides 
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+procedures to manipulate the structure of the hash map:
+`init`, `map_entry`, `rehash`, `remove_entry`, and
+`set_other_data`. It provides procedures to inquire about entries in
+the hash map: `get_other_data`, `in_map`, `unmap`.and `valid_index`.
```
```suggestion
the hash map: `get_other_data`, `in_map`, `unmap` and `valid_index`.
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+The module defines several categories of public constants. Some are
+used to parameterize the empirical slot expansion code. Others
+parameterize the slots table size, Some are used to define
```
```suggestion
parameterize the slots table size. Some are used to define
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+The constants `int_calls`, `int_depth`, `int_index`, and `int_probes`
+are used to define integer kind values for various contexts. The
+number of calls are reported and stored in entities of kind
+`int_calls`. Currently `int_calls` has the value of `INT64`. The
```
```suggestion
`int_calls`. Currently `int_calls` has the value of `int64`. The
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+`int_calls`. Currently `int_calls` has the value of `INT64`. The
+total depth, the number of inquiries needed to access all elements
+of the table, is reported and stored in entities of kind
+`int_depth`. Currently `int_depth` has the value of `INT64`. The
```
```suggestion
`int_depth`. Currently `int_depth` has the value of `int64`. The
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+of the table, is reported and stored in entities of kind
+`int_depth`. Currently `int_depth` has the value of `INT64`. The
+number of entries in the table, is reported and stored in entities of
+kind `int_index`. Currently `int_index` has the value of `INT32`.
```
```suggestion
kind `int_index`. Currently `int_index` has the value of `int32`.
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+Finally the error codes `success`, `alloc_fault`, and
+`array_size_error` are used to report the error status of certain
+procedure calla. The `succes` code indicates that no problems were
```
```suggestion
procedure calls. The `succes` code indicates that no problems were
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+### The `stdlib_hashmap_chaining` module's derived types
+
+The `stdlib_hashmap_chaining` module defines several derived
+types. The only public type is the `chaining_hash_map_type`. There are
```
```suggestion
types. The only public type is the `chaining_hashmap_type`. There are
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
````diff
@@ -0,0 +1,3112 @@
+            next => null() ! Next bucket
+    end type chaining_map_entry_type
+```
+Currently the `INT_HASH` and `INT_INDEX` have the value of `INT32`.
````
```suggestion

Currently the `int_hash` and `int_index` have the value of `int32`.
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+#### The `chaining_map_entry_ptr` derived type
+
+The type `chaining_map_entry_ptr` are used to define the elements of
```
```suggestion
The type `chaining_map_entry_ptr` is used to define the elements of
```

wclodius2 replied:
Fixed at both locations.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
````diff
@@ -0,0 +1,3112 @@
+
+```fortran
+    type :: chaining_map_entry_pool
+    ! Type inplementing a pool of allocated
````
```suggestion
    ! Type implementing a pool of allocated
```

wclodius2 replied:
Fixed both locations

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+      type(chaining_hash_map_type) :: map
+      type(int_calls) :: initial_calls
+      call init( map, fnv_1_hasher )
+      initisl_calls = calls (map)
```
```suggestion
      initial_calls = calls (map)
```

wclodius2 replied:
Fixed both locations.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Result character
+
+The result will be an integer of kind `INT_INDEX`.
```
```suggestion
The result will be an integer of kind `int_index`.
```

wclodius2 replied:
done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+      type(chaining_hash_map_type) :: map
+      type(int_index) :: initial_entries
+      call init( map, fnv_1_hasher )
+      initisl_entries = entries (map)
```
```suggestion
      initial_entries = entries (map)
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Arguments
+
+`map`): shall be a scalar variable of type
```
```suggestion
`map`: shall be a scalar variable of type
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Example
+
+    program demo_init
```
````suggestion
```fortran
    program demo_init
````

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+        type(chaining_map_type) :: map
+        call init( map,           &
+                   fnv_1a,        &
+                   slots_bits=10 )
```
````suggestion
                   slots_bits=10 )
```
````

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+##### Result value
+
+The result will be the ratio of the number of entries relative to the
+number of slots in the hash map.?
```
```suggestion
number of slots in the hash map.
```

wclodius2 replied:
Done at both locations.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Arguments
+
+`map`: shall be a scalar expression of type `chaining_hash_map_type.
```
```suggestion
`map`: shall be a scalar expression of type `chaining_hash_map_type`.
```

wclodius2 replied:
Fixed the two instances.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+      type(open_hash_map_type) :: map
+      type(int_calls) :: initial_calls
+      call init( map, fnv_1_hasher )
+      initisl_calls = calls (map)
```
```suggestion
      initial_calls = calls (map)
```

wclodius2 replied:
Fixed the two instances.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+      type(open_hash_map_type) :: map
+      type(int_index) :: initial_entries
+      call init( map, fnv_1_hasher )
+      initisl_entries = entries (map)
```
```suggestion
      initial_entries = entries (map)
```

wclodius2 replied:
Done.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Arguments
+
+`map`): shall be a scalar variable of type
```
```suggestion
`map`: shall be a scalar variable of type
```

wclodius2 replied:
Fixed the two instances.

jvdp1 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+      type(open_hash_map_type) :: map
+      real :: ratio
+      call init( map, fnv_1_hasher )
+      ratio = relative loading (map)
```
```suggestion
      ratio = loading (map)
```

wclodius2 replied:
Should be `ratio = relative_loading`.

jvdp1 reviewed (commented):
Thank you @wclodius2. Here are mainly small comments and typos. I will try to focus more on the API next week.

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Example
+
+    program demo_rehash
```
It also needs the closing ```. Done

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Example
+
+    program demo_set_other_data
```
I did a search for program  and added the Fortran code marking where missing.

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+
+##### Arguments
+
+`map`: shall be a scalar expression of type `open_hash_map_type.
```
fixed the two instances.

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+is an `intent(in)` argument. It is the non-zero index to the inverse
+table identifying the unmapped entry.
+
+* If ``inmap` is zero or `key` hass been eliminated from the table
```
Done

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+  subsequent to the generation of `inmap`, `other` is undefined.
+
+`key`: shall be a variable of type `key_type`
+`INT8`, or an allocatable length default character. It is an
```
Changed all instances of `INT8`, `INT32`, and `INT64` to `int8`, `int32`, and `int64` respectively.

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+The Fortran Standard Library provides three modules for the
+implementation of simple hash maps. These maps only accept hash
+functions with a single argument, the key, and that yield a 32 bit
+hash code, The modules will need to be modified to use hash functions
```
I have greatly revised that paragraph.

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+`stdlib_hash_32bit`, and provides wrappers to some of the
+hash functions so that they no longer need to be supplied seeds. The
+module `stdlib_hashmap_chaining` defines a datatype,
+`chaining_hash_map_type`, implementing a simple separate chaining hash
```
I changed all  mentions of `hash_map` to `hash map`.

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+module `stdlib_hashmap_chaining` defines a datatype,
+`chaining_hash_map_type`, implementing a simple separate chaining hash
+map noted more for its diagnostics than its performance. Finally the
+module, `stdlib_hashmap_open` defines a datatype,
```
Fixed

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+addressing, respectively, to deal with hash index collisions, and are
+largely defined in the separated submodules, `stdlib_hashmap_chainings`
+and `stdlib_hashmap_opens`, respectively.
+In `chaining_hash_map_type` the colliding indices are handled by using
```
Fixed see above.

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+and `stdlib_hashmap_opens`, respectively.
+In `chaining_hash_map_type` the colliding indices are handled by using
+linked lists with their roots at the hash index.
+In `open_hash_map_type`, the colliding indices are handled by searching
```
See above.

wclodius2 commented on doc/specs/stdlib_hash_maps.md:
```diff
@@ -0,0 +1,3112 @@
+The maps can take entry keys of type `key_type`.
+Both maps allow the addition and lookup of entries, and the inclusion
+of data in addition to the entry key.
+The `chaining_hash_map_type` also allows the selective removal of
```
See above

wclodius2 commented:
At this point there are several significant API related issues that I have not addressed:
1. Should there be a parent class that `chaining_hashmap_type`and open_hashmap_type`inherit from`?
2. Should the procedures be type bound?
3. I don't see any use for a general purpose container type, but I will be open to other arguments.
4. Is the use of the final routines appropriate?
5. How do I do "a link to the page of the `fibonacci_hash`"?
6. Should `equal_keys` be replaced by a defined equality?
7. Should `inmap` or `key` be the main argument for many of the functions?
8. Should `copy_key` and `copy_other` be replaced by defined assignment?

Is there any other such issue that's not bee brought up or I have not recognized?

wclodius2 commented:
Can someone mark this PR as needing reviewers and on the topic of containers?

wclodius2 commented:
I forgot one other issue with the API.
11. Should the `other_type` be a wrapper for a `CLASS(*)` internal component?

jvdp1 commented:
> Can someone mark this PR as needing reviewers and on the topic of containers?

Done.

jvdp1 commented:
Here are some first answers:

> 2. Should the procedures be type bound?

I think they should be both. However, the first implementation could be as it is now.

> 4. Is the use of the final routines appropriate?

I think so.

> 5. How do I do "a link to the page of the `fibonacci_hash`"?

It should be a static link such as "[fibonacci_hash](..page/specs/stdlib_hash_procedures.html#fibonacci_hash-maps-an-integer-to-a-smaller-number-of-bits)" with the link as (..page/specs/stdlib_hash_procedures.html#fibonacci_hash-maps-an-integer-to-a-smaller-number-of-bits)

LKedward reviewed (commented):
Thanks for this @wclodius2, this will make a good addition to stdlib.

The underlying implementation, as implied by the spec, seems good to me in general,
though I agree with others that this would be well-suited to having a base class for the two hash maps. This would also remove redundancy in the documentation for both hash map types.

> 1. Should there be a parent class that chaining_hashmap_type and open_hashmap_type inherit from?

Unless there has been a discussion on avoiding inheritence in stdlib, I think it makes sense to use it here.

> 2. Should the procedures be type bound?

Yes, this simplifies the `use` statement and is intuitive for containers.

> 3. I don't see any use for a general purpose container type, but I will be open to other arguments.
> 11. Should the other_type be a wrapper for a CLASS(*) internal component?

While I'm not a heavy user of hash tables, it is useful to be able to store arbitrary types.
With that said, it isn't a necessity and using `class(*)` does come with a performance degradation.
(My own hash table implementation allows storing any intrinsic scalar or
[derived type](https://lkedward.github.io/fhash/page/2-derived-type-demo/index.html) using `class(*)`, and this does incur a runtime penalty).
I'm interested to hear what others think about this.

> 4. Is the use of the final routines appropriate?

This seems fine to me as described.

> 6. Should equal_keys be replaced by a defined equality?

Yes, I don't see any reason not to do so.

> 7. Should inmap or key be the main argument for many of the functions?

The use of `inmap` as an intermediary is unusual for a hash map based on my (limited) experience.
It clearly offers a performance advantage for the case described, but it is a little long-winded (key->inmap->value) compared to a typical hash map usage (key->value).
Is it possible to have both options (`inmap` and `key`) using a generic interface?

What is the danger of returning a pointer instead as suggested by @awvwgk?
(Again, my own hash table allows retreiving either a value or a pointer.)

> 8. Should copy_key and copy_other be replaced by defined assignment?

I have been caught out by bugs that easily and silently creep in when using defined assignments.
(See also discussion on Discourse: https://fortran-lang.discourse.group/t/should-we-avoid-assignment-of-derived-types-in-robust-programs/2359/5)
IMO, if it's not absolutely necessary (_e.g._ for reference counting) then I would avoid using defined assignment.

wclodius2 commented:
I have significantly reduced the discussion of `fibonacci_hash` and have replaced it with a link to `.../stdlib_hash_functions.html`. Let me know if it should be further changed.

wclodius2 commented:
I will change the `other_type` so that its component ia a `CLASS(*)` variable.. Is it possible to do equality tests on such variables?

gareth-nx commented:
> I will change the other_type so that its component ia a CLASS(*) variable.. Is it possible to do equality tests on such variables?

I think it needs `select_type` (?). 

It would be good to understand whether this leads to performance degredation, and how much.

wclodius2 commented:
I don't think that making `other_type` into a wrapper for a allocatable `class(*) `component has significantly more overhead than making it a wrapper for an allocatable `int8` vector, and the `class(*)` component is simpler to use, so I have changed its description to that of a `class(*)` allocatable.

awvwgk commented:
`class(*)` cannot reliably tested for equality, I actually don't think this has to concern the hash map, since it supposed to only store the data and not touch it.

A problem I can see with the current byte representation is that it might not be able to represent any data type, *i.e.* I'm unsure whether it is a good idea to `transfer` a derived type with procedure pointers or a polymorphic object into bytes to store than. Reconstruction will be unsafe while `class(*)` provides at least a mechanism to retrieve this information via `select type` again.

If `class(*)` is an issue for the user, it might be possible to just store the data via `c_loc` as `type(c_ptr)` and than convert it back with `c_f_pointer`. This way the hash map does not have to own the data.

wclodius2 commented:
I don't know why I asked about equality tests for the other data. It is not an issue at all.

wclodius2 commented:
I have decided that the repeating documentation for the two hash maps is a pain that will be reduced by using an inheritance hierarchy, and that type  bound procedures would be more intuitive to users. This leads to the following changes.
1. There will be a parent class that `chaining_hashmap_type` and `open_hashmap_type` will inherit from.
2. That the parent class and the structure of the child classes will be defined in a module `stdlib_hashmaps` with the obvious filename.
3. That the procedures for the classes will be type bound.
4. That the procedures specific to `chaining_hashmap_type` and `open_hashmap_type` will be implemented in the submodules `stdlib_hashmap_chaining` and `stdlib_hashmap_open`, respectively, with the obvious filenames.

Making these changes in the documentation and the codes will take a number of days, probably more than a week, maybe more than two weeks.

awvwgk commented:
Thanks, William. I have a third kind of implementation of a hash map I want to propose later ([cuckoo hash](https://en.wikipedia.org/wiki/Cuckoo_hashing)), if you have a base class ready, I would like to try it by porting my implementation.

aradi commented:
I am lacking the time to make a proper review, but I could not resist to make few general comments about containers, just in case you find it useful:

* One should not rely on array of integers (or characters) as data type, especially not for the value. Transfering anything else than intrinsic types via `transfer()` to integers and back can lead to unpredictable behaviour IMO. Think about a derived type with allocated pointer components. You transfer it to integer representation and put in in the hash-map. What does the map contain? The full data of the type instance? Or only the pointer descriptors? And what happens if the original type instance is finalized, and then you read out the integer representation from the hash-map and transfer it back? Would the pointers within the retrieved instance point to something meaningful? (likely not). Using `class(*)` instead is more robust, although with run-time penalties and the annoying `select type` everywhere. Users would have to write wrappers for all their custom types they would like to put into the container. `c_loc()` is basically like transfer, in the sense, that you do not get compile-time (and possible not even immediate run-time errors), when converting back the retrieved item to the wrong data type.

* I think, making internal details of the implementation of the map (like an integer index) of the public API would constrain future changes of the map-implementation. Also, how long is the integer index valid? Would it change, in case the hash table must grow? I think, returning a pointer to the data (as proposed above) is somewhat more robust, but it should be still well documented, whether there are any map operations, which make the pointer (returned earlier) invalid.

* Unfortunately, whatever you do, you will not be able to fix the biggest problem of Fortran: the lack of templates/generic programming. If you wanted robust, efficient and easy to use containers with **compile-time** errors for many easy mistakes (like putting in an integer and reading it out as a real), you would need container templates (macros), which the user can instantiate with the actual type. There are already examples using cpp (see https://github.com/SCM-NV/ftl), and one can do it even much nicer with fypp :wink: (see https://github.com/dftbplus/dftbplus/blob/master/src/dftbp/include/allocatablelist.fypp for a simple list template and https://github.com/dftbplus/dftbplus/blob/master/src/dftbp/dftb/repulsive/repulsivelist.F90 for an instantiation with a complex derived type). But, unfortunately, that would require the consumers of stdlib to use fypp, which is against the design principle of stdlib. :disappointed:

wclodius2 commented:
@aradi
My response:
* While I have retained integer vectors and character strings for the `key`, I am now using `class(*)` for the associated data. This allows the associated data to be almost anything you want. However the `key` needs to be hashed, and, in the event of hash collisions, it needs a test for equality. All general purpose hash functions, I have found, ultimately rely on being able to map the thing being hashed to an vector of (unsigned) eight bit integers, typically, in C related languages, by a typecast. This representation also allows a straightforward equality test. The closest thing to a typecast, in Fortran, is `transfer`. Fortunately you almost invariably want the `key` to be a straightforward "value" type: a scalar of intrinsic type, a vector of intrinsic types, a character string, or a derived type whose components' types are "value" types. For most of these a transfer to an `int8` vector is straightforward. For the derived type it will have to have an associated function that maps, using transfer, the different component to different portions of an `int8` vector. For most of the these it is easy to devise a straightforward inverse transfer that takes the `int8` vector and generates a value of the original type. It is more complicated, but often possible, to define a function that uses transfer to generate the appropriate derived type value from the `int8` vector. I don't know of a way, in Fortran, to deal with keys that are pointers or derived types with pointer components. Those who want that are out of luck.
* I will probably replace the use of `inmap` with `key`, but I want  to complete the transformation to a parent type and its extensions first. FWIW I have already converted the code and its tests, but documenting the changed code will still take a number of days.
* I cannot fix what cannot be fixed.

wclodius2 commented:
I have oushed a new version of `stdlib_hash_maps.md`. The documented changes include
1. Changing `equal_keys` to an equality operator, `==`.
2. Removing the procedure `relative_loading` as too specialized (useful only for open hash maps), and probably of no interest to the vast majority of users.
3. Converting the hash map types to a class based system with the associated changes:
    * Created the new module file `stdlib_hashmaps.f90` with the module `stdlib_hashmaps`;
    * Have `stdlib_hashmaps` define three public types: `hashmap_type`, an abstract type to serve as the parent type for hash map implementations, `chaining_hashmap_type` that extends `hashmap_type` to implement a separate chaining hash map, and `open_hashmap_type` that extends `hashmap_type` to implement an linear open addressing hash map;
    * Converted the modules `stdlib_hashmap_chaining` and `stdlib_hashmap_open` to submodules of `stdlib_hashmaps` that implement the procedures of the subtypes `chaining_hashmap_type` and `open_hashmap_type`;
    * Have made the procedures of the types class bound.

Overall I am pleased with the conversion to a class based type hierarchy, and am very pleased with making the procedures type bound. It greatly simplifies the use of the modules.

Now comes the question as to whether the `inmap` index should be replaced with the `key`, and how that would affect the API. The `inmap` index has the advantage that multiple references to the same entry can use a single search so it is sometimes more efficient than relying on the `key`. An `inmap` value of 0 also serves as an error flag, which, in the absence of exception handling, would have to be replaced by a status or error flag dummy argument. It has the disadvantage that it exposes the implementation making it difficult to implement other hash maps that extend `hashmap_type`, will be out of date if the corresponding entry is deleted from the map, and may be more wordy than a `key` based entry access. Overall I am inclined to think that using a `key` instead of the `inmap` index is to be preferred.

If the `inmap` index is to be replaced with the `key` then we should consider the details of the changes in the API. Currently the detailed API is determined by the components of the abstract `hashmap_type`.  These are given by the type specification:

```fortran
    type, abstract :: hashmap_type
!! Version: Experimental
!!
!! Type implementing an abstract hash map
        private
        integer(int_calls) :: call_count = 0
!! Number of calls
        integer(int_calls) :: probe_count = 0
!! Number of probes since last expansion
        integer(int_calls) :: total_probes = 0
!! Cumulative number of probes
        integer(int_index) :: num_entries = 0
!! Number of entries
        integer(int_index) :: num_free = 0
!! Number of elements in the free_list
        integer(int32)     :: nbits = default_bits
!! Number of bits used to address the slots
        procedure(hasher_fun), pointer, nopass :: hasher => fnv_1_hasher
!! Hash function
    contains
        procedure, non_overridable, pass(map) :: calls
        procedure, non_overridable, pass(map) :: entries
        procedure, non_overridable, pass(map) :: map_probes
        procedure, non_overridable, pass(map) :: num_slots
        procedure, non_overridable, pass(map) :: slots_bits
        procedure(get_other), deferred, pass(map)    :: get_other_data
        procedure(in_map), deferred, pass(map)       :: in_map
        procedure(init_map), deferred, pass(map)     :: init
        procedure(loading), deferred, pass(map)      :: loading
        procedure(map_entry), deferred, pass(map)    :: map_entry
        procedure(rehash_map), deferred, pass(map)   :: rehash
        procedure(remove_entry), deferred, pass(map) :: remove
        procedure(set_other), deferred, pass(map)    :: set_other_data
        procedure(total_depth), deferred, pass(map)  :: total_depth
        procedure(unmap), deferred, pass(map)        :: unmap
        procedure(valid_index), deferred, pass(map)  :: valid_index
    end type hashmap_type
```

Of these procedures only `get_other_data`, `in_map`, `map_data`, `remove`, `set_other_data`, `unmap`, and `valid_index` have an `inmap` argument. The procedure `unmap` is meaningless without a public `inmap` and is not used internally by any of the procedures so it should  be removed entirely. The procedure `in_map` is not useful publicly, but is useful to map a `key` argument the maps' internal structure so should be retained as a private procedure. The `get_other_data` procedure should have its `inmap` argument replaced by a `key` argument, and a status/flag argument, say `exists`, should be added.be added.  The `map_entry` procedure should have its `inmap` argument replaced by a `key` argument and a flag/status argument, say `duplicate`, should be added.  The `set_other_data` procedure should have its `inmap` argument replaced by a `key` argument and a flag/status argument, say `not_present`, should be added. The `valid_index` procedure with an `inmap` argument, should be replaced by a `valid_key` procedure with a `key` argument.

wclodius2 commented:
I have added a new markdown document, `stdlib_hashmaps.md`, that documents my current proposed API for hash map implementations that use `key` and not `inmap` for map entry, retrieval, and modification. I have retained the old `stdlib_hash_maps.md` so comparisons of the two API's can be made directly. I also fixed some typos in `stdlib_hash_maps.md`.

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+
+##### Description
+
+Places the data from `value` in a `key_type` or an `other_type`.
```
Could you provide some details on how the key is "placed" in value?

wclodius2 replied:
I don't think the details are necessary for the documentation, butter your information. For a key the internal `int8` array is allocated to the proper size, then for an `int8` value the contents are copied, while for a character string value the contents are transferred. For other data, the contents are allocated with the value as the source.

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+`other`: shall be a scalar variable of type `other_type`. It
+is an `intent(out)` argument.
+
+`value`: if the first argument is `key` `vaalue` shall be a default
```
```suggestion
`value`: if the first argument is `key` `value` shall be a default
```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+types using separate chaining hashing and open addressing hashing. The
+derived type `hashmap_type` is the parent type to its two
+extensions: `chaining_hashmap_type` and `open_hashmap_type`.
+`chaining_hashmap_type`. The extension types provide 
```
```suggestion
The extension types provide 
```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+
+The constant `load_factor` is only used by the `open_hashmap_type`. It
+specifies the maximum fraction of the available slots that may be
+filled before expansion occurs. The current `load_factor = ).5625` so
```
```suggestion
filled before expansion occurs. The current `load_factor = 0.5625` so
```

jvdp1 replied:
Why is it this value? How was it derived?

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+four private derived types, `open_map_entry_type`,
+`open_map_entry_list`, `open_map_entry_ptr`, and `open_map_entry_pool`
+are used in the implementation of the `open_hashmap_type` public
+type:. Each of these types are described below. 
```
```suggestion
type. Each of these types are described below. 
```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+`intent(out)` argument.  If `true` an entry with the given `key`
+exists in the map, if false `other` is undefined.
+
+* The following is an example of the retrieval of other data
+  associated with a `key`:
+
+
+##### Example
+
```
```suggestion
##### Example

 The following is an example of the retrieval of other data
  associated with a `key`:



```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+        call map % init( fnv_1_hasher )
+        call set( key, [ 0_int8, 1_int8, 2_int8, 3_int8, 4_int8 ] )
+        call set( other, data )
+        call map % map_entry( key, other. conflict )
```
```suggestion
        call map % map_entry( key, other, conflict )
```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+        if ( .not. conflict ) then
+            call map % get_other_data( key, other )
+        else
+            stop 'Key is already present in the map.''
```
```suggestion
            stop 'Key is already present in the map.'
```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+
+##### Syntax
+
+`call [[stdlib_hashmaps:map%init]]( hasher [, slots_bits, status ] ] )`
```
```suggestion
`call [[stdlib_hashmaps:map%init]]( hasher [, slots_bits, status ] )`
```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+##### Argument
+
+`map` (pass) - shall be an expression of class `chaining_hashmap_type`
+or ``open_hashmap_type`. It is an `intent(in)` argument.
```
```suggestion
or `open_hashmap_type`. It is an `intent(in)` argument.
```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+
+`existed` (optional): shall be a scalar variable of type default
+logical. It is an `intent(out)` argument. If present with the value
+`true` the entry existed 
+in the map before removal, if false the entry was not present to be
```
```suggestion
`true` the entry existed 
in the map before removal, if `false` the entry was not present to be
```

jvdp1 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+##### Argument
+
+`map` (pass): shall be a scalar expression of class
+`chaining_hashmap_type`. It is an `intent(in)` argument. It is the
```
```suggestion
`hashmap_type`. It is an `intent(in)` argument. It is the
```

jvdp1 reviewed (commented):
Thank you for this new version. I believe that comments from @awvwgk and @aradi were integrated in this version.
On overall, I think that the API look good.

I am wondering if the key DT could be directly used by the user (to avoid to call `set`

wclodius2 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+
+# The `stdlib_hashmap_wrappers`, `stdlib_hashmap_chaining` and `stdlib_hashmap_open` modules
+
+(TOC)
```
Done

wclodius2 commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2104 @@
+
+### The `stdlib_hashmap_wrappers`'s constant, `int_hash`
+
+The constant `int_hash` is used to define the integer kind value for
```
It is imported from `stdlib_hash_32bit`. I will document that.

wclodius2 commented:
In pushing the changes suggested by @jvdp1, and additional fixes found while addressing his concerns, I am having massive PR check failures. All the failures seem to be associated with the message "Warning: Failed to download action " which suggests to me that it is either a network problem, or a problem caused by pushing changes too rapidly. I will try pushing a small change tomorrow and see what happens.

wclodius2 commented:
Its working now.

awvwgk commented:
We have two spec files now `stdlib_hash_maps.md` and `stdlib_hashmaps.md`. I guess one is a duplicate.

jvdp1 commented:
> . I have retained the old `stdlib_hash_maps.md` so comparisons of the two API's can be made directly. I also fixed some typos in `stdlib_hash_maps.md`.

A bit confusing indeed, but @wclodius2 [wrote the following](https://github.com/fortran-lang/stdlib/pull/611#issuecomment-1009281322):
```
I have added a new markdown document, stdlib_hashmaps.md
....
I have retained the old stdlib_hash_maps.md so comparisons of the two API's can be made directly. I also fixed some typos in stdlib_hash_maps.md.
```

awvwgk commented:
Okay, seems a bit complicated. I will give the new one a look soon.

jvdp1 commented:
> Okay, seems a bit complicated. I will give the new one a look soon.

`stdlib_hashmaps.md` is a copy of `stdlib_hash_maps.md` updated with all reviewers' comments. Therefore, only `stdlib_hashmaps.md` should be read.

wclodius2 commented:
They are not copies of one another. `stdlib_hashmaps.md` greatly modifies `stdlib_hash_maps.md` to use `key` instead of `inmap` to access records.

wclodius2 commented:
In trying to implement the API described in the first version of `stdlib_hashmaps.md` I discovered that for some of the procedures the hash map had to have `intent(inout)` so `valid_key` had to be a subroutine rather than a function. I have renamed `valid_key` to `key_test`, made the argument `map` to `get_other_dats` have `intent(inout)`, and have fixed a small typo. `stdlib_hashmaps.md` has been updated to reflect these changes.
The proposed API that uses `key` to access entries has been tested for performance and self consistency and can be compared with the API of `stdlib_hash_maps.md` that uses the `inmate` index to access entries. On a MacBook Pro with an Apple M1 Pro processor running MacOS Monterrey 12.1, with gfortran 11.2 as the compiler the following performance is found:

For a chaining hash map using `inmap` and two hashing algorithms

|         Algorithm |      Process |        Data Set |   Time (s) |
|             FNV-1 |   Enter data |   16 byte words |    0.05254 |
|             FNV-1 | Inquire data |   16 byte words |    0.01735 |
|             FNV-1 |     Get data |   16 byte words |    0.02447 |
|             FNV-1 |  Rehash data |   16 byte words |    0.00966 |
|             FNV-1 |  Remove data |   16 byte words |    0.01974 |
|             FNV-1 |   Enter data |  256 byte words |    0.01184 |
|             FNV-1 | Inquire data |  256 byte words |    0.00972 |
|             FNV-1 |     Get data |  256 byte words |    0.01021 |
|             FNV-1 |  Rehash data |  256 byte words |    0.00810 |
|             FNV-1 |  Remove data |  256 byte words |    0.00991 |
|      Seeded_Water |   Enter data |   16 byte words |    0.03489 |
|      Seeded_Water | Inquire data |   16 byte words |    0.01385 |
|      Seeded_Water |     Get data |   16 byte words |    0.02202 |
|      Seeded_Water |  Rehash data |   16 byte words |    0.00688 |
|      Seeded_Water |  Remove data |   16 byte words |    0.01745 |
|      Seeded_Water |   Enter data |  256 byte words |    0.00794 |
|      Seeded_Water | Inquire data |  256 byte words |    0.00581 |
|      Seeded_Water |     Get data |  256 byte words |    0.00627 |
|      Seeded_Water |  Rehash data |  256 byte words |    0.00411 |
|      Seeded_Water |  Remove data |  256 byte words |    0.00597 |

For an open hash map using `inmap` and two hashing algorithms

|         Algorithm |      Process |        Data Set |   Time (s) |
|             FNV-1 |   Enter data |   16 byte words |    0.05024 |
|             FNV-1 | Inquire data |   16 byte words |    0.01513 |
|             FNV-1 |     Get data |   16 byte words |    0.02200 |
|             FNV-1 |  Rehash data |   16 byte words |    0.01011 |
|             FNV-1 |  Remove data |   16 byte words |    0.01671 |
|             FNV-1 |   Enter data |  256 byte words |    0.01186 |
|             FNV-1 | Inquire data |  256 byte words |    0.00971 |
|             FNV-1 |     Get data |  256 byte words |    0.01018 |
|             FNV-1 |  Rehash data |  256 byte words |    0.00813 |
|             FNV-1 |  Remove data |  256 byte words |    0.00985 |
|      Seeded_Water |   Enter data |   16 byte words |    0.03312 |
|      Seeded_Water | Inquire data |   16 byte words |    0.01178 |
|      Seeded_Water |     Get data |   16 byte words |    0.01977 |
|      Seeded_Water |  Rehash data |   16 byte words |    0.00709 |
|      Seeded_Water |  Remove data |   16 byte words |    0.01456 |
|      Seeded_Water |   Enter data |  256 byte words |    0.00799 |
|      Seeded_Water | Inquire data |  256 byte words |    0.00574 |
|      Seeded_Water |     Get data |  256 byte words |    0.00622 |
|      Seeded_Water |  Rehash data |  256 byte words |    0.00411 |
|      Seeded_Water |  Remove data |  256 byte words |    0.00591 |

For a chaining hash map using `key` and two hashing algorithms

|         Algorithm |      Process |        Data Set |   Time (s) |
|             FNV-1 |   Enter data |   16 byte words |    0.04585 |
|             FNV-1 | Inquire data |   16 byte words |    0.01552 |
|             FNV-1 |     Get data |   16 byte words |    0.02299 |
|             FNV-1 |  Rehash data |   16 byte words |    0.00974 |
|             FNV-1 |  Remove data |   16 byte words |    0.01792 |
|             FNV-1 |   Enter data |  256 byte words |    0.01186 |
|             FNV-1 | Inquire data |  256 byte words |    0.00972 |
|             FNV-1 |     Get data |  256 byte words |    0.01021 |
|             FNV-1 |  Rehash data |  256 byte words |    0.00810 |
|             FNV-1 |  Remove data |  256 byte words |    0.00989 |
|      Seeded_Water |   Enter data |   16 byte words |    0.03481 |
|      Seeded_Water | Inquire data |   16 byte words |    0.01414 |
|      Seeded_Water |     Get data |   16 byte words |    0.02198 |
|      Seeded_Water |  Rehash data |   16 byte words |    0.00680 |
|      Seeded_Water |  Remove data |   16 byte words |    0.01765 |
|      Seeded_Water |   Enter data |  256 byte words |    0.00793 |
|      Seeded_Water | Inquire data |  256 byte words |    0.00585 |
|      Seeded_Water |     Get data |  256 byte words |    0.00625 |
|      Seeded_Water |  Rehash data |  256 byte words |    0.00410 |
|      Seeded_Water |  Remove data |  256 byte words |    0.00600 |

For an open hash map using `key` and two hashing algorithms

|         Algorithm |      Process |        Data Set |   Time (s) |
|             FNV-1 |   Enter data |   16 byte words |    0.04995 |
|             FNV-1 | Inquire data |   16 byte words |    0.01509 |
|             FNV-1 |     Get data |   16 byte words |    0.02197 |
|             FNV-1 |  Rehash data |   16 byte words |    0.01008 |
|             FNV-1 |  Remove data |   16 byte words |    0.01665 |
|             FNV-1 |   Enter data |  256 byte words |    0.01178 |
|             FNV-1 | Inquire data |  256 byte words |    0.00969 |
|             FNV-1 |     Get data |  256 byte words |    0.01019 |
|             FNV-1 |  Rehash data |  256 byte words |    0.00812 |
|             FNV-1 |  Remove data |  256 byte words |    0.00985 |
|      Seeded_Water |   Enter data |   16 byte words |    0.03315 |
|      Seeded_Water | Inquire data |   16 byte words |    0.01229 |
|      Seeded_Water |     Get data |   16 byte words |    0.01997 |
|      Seeded_Water |  Rehash data |   16 byte words |    0.00705 |
|      Seeded_Water |  Remove data |   16 byte words |    0.01472 |
|      Seeded_Water |   Enter data |  256 byte words |    0.00795 |
|      Seeded_Water | Inquire data |  256 byte words |    0.00576 |
|      Seeded_Water |     Get data |  256 byte words |    0.00622 |
|      Seeded_Water |  Rehash data |  256 byte words |    0.00411 |
|      Seeded_Water |  Remove data |  256 byte words |    0.00592 |

jvdp1 commented:
@wclodius2 thank you for the update and performance tests. I am sorry to come back so late!
I have no additional comments on the API. So I guess that at this stage you could submit the source code, when you are ready, right?

wclodius2 commented:
I have decided:

1. that the inmap index was no more efficient for most purposes than the key; and
2. that the title `stdlib_hashmaps.md` was more consistent with the rest of the documentation than the title `stdlib_hash_maps..md`.

As a result I have removed `stdlib_hash_maps.md` from the distribution, updated the `index.md` to reflect that change, and pushed the result. I now plan to add the implementation of the codes described in `stdlib_hashmaps.md`

jvdp1 reviewed (commented):
Thank you for the PR.
Finally I have some times for stdlib.

Before continuing to review this PR, why are the procedures `set` and `get` subroutines, instead of functions

wclodius2 commented:
`set`  is a subroutine rather than a function because the `map` argument  is `intent(inout)`. It is `intent(inout)` because `set` modifies a small part of the hashmap structure, rather than duplicating almost the whole structure as a new object.

`get` is a subroutine rather than a function because it has the optional `exists` error flag as an `intent(inout)` argument. A function with an `intent(out)` argument has undefined behavior.

wclodius2 commented on src/stdlib_hashmap_wrappers.f90:
```diff
@@ -0,0 +1,378 @@
+    end subroutine set_other
+
+
+    subroutine set_int8_key( key, value )
```
Currently the user should pass the equivalent of `transfer(real_array, 0_int8, 4*size(real_array)`.

14NGiestas replied:
Is that user-documented? I was wondering about the usage of such int8 procedure and it seems that is to achieve a wild generic keying, if so some usage examples in the docstring would be appreciated, I don't know if it's subject to change but it should mention the `transfer` at least for now.

I'm passing basically anything as a key and it feels really consistent:
```f90
    real(dp) :: real_thing(100)
    integer(int8), allocatable :: key_payload(:)
    real_thing = -137.0
    key_payload = transfer(real_thing, 0_int8, sizeof(real_thing))
    call set(key, key_payload)
    ! Trying to get a string why not
    call get(key, iwantmystring)
    print*, iwantmystring ! it prints a lot of " a" printed, which is ok
    ! Trying to get a real_array
    call get(key, key_payload)
    print*, transfer(key_payload, 0.0_dp, 100) ! A hundred -137  as expected
```

14NGiestas commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2088 @@
+      implicit none
+      type(chaining_hashmap_type) :: map
+      type(int_calls) :: initial_calls
+      call map % init(fnv_1_hasher )
```
```suggestion
      call map % init( fnv_1_hasher )
```
Just whitespace that feels odd, maybe it's time to discuss a CI enforced auto-style like  [fprettify](https://github.com/pseewald/fprettify) #3 and ways to avoid erroneous docs examples #639, so we can stop worrying about them.

14NGiestas commented on src/stdlib_hashmap_chaining.f90:
```diff
@@ -0,0 +1,848 @@
+        integer, intent(in), optional              :: slots_bits
+        integer(int32), intent(out), optional      :: status
+
+        character(256)          :: errmsg = ''
```
It will, but since the program is terminated when this is used, it won't matter much.
Anyway, I don't see why this can't be changed (only if the `allocate` leave the `errmsg` unchanged)

```suggestion
        character(256)          :: errmsg
```

14NGiestas commented on src/stdlib_hashmap_open.f90:
```diff
@@ -0,0 +1,875 @@
+                map % free_list => free_list % next
+                free_list % target => null()
+                free_list % next => null()
+                if (bucket % inmap == 0) stop "bucket % inmap == 0"
```
Unhandled fatal error `stop` will terminate the program, should this really be fatal?.

wclodius2 replied:
If I have coded correctly the only way you can get `bucket % inmap == 0` is if there has been bit errors in either the address of `bucket`, the pointer to `inmap`, or the value of `inmap`.

14NGiestas commented on src/stdlib_hashmap_open.f90:
```diff
@@ -0,0 +1,875 @@
+                     size( map % inverse, kind=int_index ) ) then
+                    call expand_inverse( map )
+                end if
+                if ( map % num_entries == 0 ) stop "MAP % NUM_ENTRIES == 0."
```
[Same as #r865894471](https://github.com/fortran-lang/stdlib/pull/611/files#r865894471)

wclodius2 replied:
This is in a section of code where we are attempting to expand the entry pool because it is too small for the number of entries. It cannot be too small if `map % num_entries == 0`, so either the logic of the rest of the code is wrong or there has been a bit error. Mind you we could detect more errors in the code if the comparison were `map % num_entries <= 0`.

wclodius2 replied:
I have changed this in my copy and will check it in shortly.

14NGiestas commented on src/stdlib_hashmap_wrappers.f90:
```diff
@@ -0,0 +1,378 @@
+            key_as_char = key_size
+        case(2)
+            if ( iand( key_size, 1_int64 ) > 0 ) then
+                stop 'KEY does not map to a character string.'
```
[Same as #r865894471](https://github.com/fortran-lang/stdlib/pull/611/files#r865894471) and the other cases around here

14NGiestas replied:
Ok, such stops can only occur if the user environment/compiler/combination-of-flags make the `character_storage_size` unusual. I guess you should place some comments here to make sure the reader is aware that this checking is happening.
I see now how the Unicode thing started.

wclodius2 replied:
This is mapping a byte array to a string. `case(2)` is when the number of bytes in a default character is two. (I believe that can occur in MS Windows.) In that case, the `key_size` can never be an odd number. Either there is a problem with the logic of the rest of the code, or a bit error has occurred. This is unrecoverable. This might be clearer by making the test `and( key_size, 1_int64 ) == 1`.

14NGiestas replied:
Then it would make sense to change the `stop` message to something less cryptic and leave some comments about how such an edge case could ever happen. This is probably like an ICE - internal compiler error, an inconsistency that is completely unrecoverable.

I can think something like:
```suggestion
                error stop 'Internal Error at stdlib_hashmaps: System uses 2 bytes per character, so key_size can't be an odd number'
```

14NGiestas replied:
If there is a consistency check it means it can be triggered somehow (incorrect code when developing for example), otherwise, it shouldn't be part of the code in the first place and compilers just optimize it off: Ex: `if (.false.) stop`.
If such a thing ever happens it would be good for anyone experiencing that to know that is something related to an inconsistent internal state and not something else (like incorrect inputs).
Does this make any sense?

wclodius2 commented:
@14NGiestas I don't know what you mean by unguarded `stop`s. The three examples you give are all guarded by `if` statements.

14NGiestas commented:
@wclodius2 I just want to make clear that such `stop` statements won't get the user by surprise and they can't recover from it by simply passing a `status`.
I will play a little more to see if I can break it :)

14NGiestas commented on src/stdlib_hashmap_wrappers.f90:
```diff
@@ -0,0 +1,378 @@
+!!     key   - the output key
+!!     value - the input CHARACTER string
+        type(key_type), intent(out) :: key
+        character(*), intent(in)    :: value(:)
```
I think this is meant to be a scalar

```suggestion
        character(*), intent(in)    :: value
```

As I tried the following:

```fortran
program main
use stdlib_hashmaps
use stdlib_hashmap_wrappers
implicit none
type(key_type) :: key
type(open_hashmap_type)   :: map
character(:), allocatable :: iwantmystring


call map % init( fnv_1_hasher, slots_bits=10 )
call set(key, ["Hello World", "What World?"])
call get(key, iwantmystring)
print*, iwantmystring ! It shows Hello World

end program
```

14NGiestas commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2088 @@
+      hasher_pointer => fnv_1a_hasher
+      array1 = [ 5_int8, 4_int8, 3_int8, 1_int8, 10_int8, 4_int8 ]
+      call set( key, array1 )
+      hash = hassher_pointer(key)
```
```suggestion
      hash = hasher_pointer(key)
```

wclodius2 replied:
I have corrected it in my version and will push it later.

ivan-pi commented on doc/specs/stdlib_hashmaps.md:
```diff
@@ -0,0 +1,2088 @@
+
+##### Arguments
+
+`key_in`: shall be a scalar expression of type `key_type`. It
+is an `intent(in)` argument.
+
+`key_out`: shall be a scalar variable of type `key_type`. It
+is an `intent(out)` argument.
```
What about `old_key` and `new_key` instead of `key_in` and `key_out`? I think it's important to pick variables names which "read" well, i.e "make a call to `copy_key`, passing an `old_key` as input, and receiving a `new_key` as output",

wclodius2 replied:
I have done this, but wonder if `key_copy` Might be better.

14NGiestas commented on src/stdlib_hashmap_wrappers.f90:
```diff
@@ -0,0 +1,378 @@
+            key_as_char = key_size
+        case(2)
+            if ( iand( key_size, 1_int64 ) > 0 ) then
+                error stop 'Internal Error at stdlib_hashmaps: System uses 2 bytes per character, so key_size can't be an odd number'
```
```suggestion
                error stop "Internal Error at stdlib_hashmaps: System uses 2 bytes per character, so key_size can't be an odd number"
```
There was a small issue with escaping in "can't"

14NGiestas commented on src/stdlib_hashmap_wrappers.f90:
```diff
@@ -0,0 +1,378 @@
+            key_as_char = key_size
+        case(2)
+            if ( iand( key_size, 1_int64 ) > 0 ) then
+                error stop "Internal Error at stdlib_hashmaps: System uses 2 bytes per character, so key_size can't be an odd number"
```
```suggestion
                error stop "Internal Error at stdlib_hashmaps:&
                              & System uses 2 bytes per character, so&
                              & key_size can't be an odd number"
```
*sights* line is probably too long

14NGiestas commented on src/stdlib_hashmap_wrappers.f90:
```diff
@@ -206,9 +209,10 @@ subroutine get_char_key( key, value )
+                           "key_size can't be an odd number"
             end if
             key_as_char = ishft( key_size, -1 )
         case(4)
```
The other cases below could benefit from the same improvements.

wclodius2 replied:
I believe I have fixed the two cases you refer to.

14NGiestas commented on src/stdlib_hashmap_chaining.f90:
```diff
@@ -0,0 +1,848 @@
+            gentry => pentry
+            map % probe_count = map % probe_count + 1
+            if (.not. associated( gentry ) ) then
+                write(error_unit,*) "gentry not associated"
```
This line feels odd, as far as I can't tell we didn't find the index and we are returning `inmap=0` but also it writes to `error_unit`, plus this subroutine is only used internally, right?

You should probably mention that this is how you are signaling a missing index in its docstring and remove this line.

wclodius2 replied:
I believe I have now addressed this issue.

wclodius2 commented:
@14NGiestas git shows that I still have one change request from you that I have not addressed. Do you know which one that is?

14NGiestas reviewed (approved):
I think you addressed all my issues with this PR, I have nothing to add so far, so I'm giving my approval.

jvdp1 commented:
@wclodius2 sorry for the delay. I opened a [PR](https://github.com/wclodius2/stdlib/pull/12) in your repo with some propositions for the specs.

wclodius2 commented:
@jvdp1 I believe I have checked in your changes.

jvdp1 commented:
> @jvdp1 I believe I have checked in your changes.

Good. I am continuing the review.

jvdp1 commented:
@wclodius2 I opened a  2 PR in your repo with some propositions for the docstrings and the tests:

* [PR 13](https://github.com/wclodius2/stdlib/pull/13) - addition of links
* [PR 14](https://github.com/wclodius2/stdlib/pull/14) - addition of tests

wclodius2 commented:
@jvdp1 I have committed PR #13 and PR #14.

jvdp1 reviewed (approved):
Thank you @wclodius2 for this feature.

jvdp1 commented:
@fortran-lang/stdlib @milancurcic  @ivan-pi @LKedward @awvwgk @gareth-nx and others interested by hash maps:
There are now 2 complete reviews. I am sure that we can still discuss the API. However, I propose that we merge this PR as `experimental` such that users can test it and play with it, except if you still  see a major issue that would prohibit its merging.

milancurcic commented:
I haven't reviewed but agree for this PR to go forward.

wclodius2 commented:
I have also updated from fortran-lang/stdlib.

jvdp1 commented:
We discussed the status of this PR during the [monthly call](https://fortran-lang.discourse.group/t/fortran-monthly-call-june-2022/3668/9?u=jeremie.vandenplas), and agreed to merge it such that users can test and play with this new feature. So, I will merge it.

@wclodius2 Thank you for this PR.
